### PR TITLE
Check if only public property is documented

### PR DESCRIPTION
**Pull request by atulgpt**
Fixes https://github.com/detekt/detekt/issues/5217

### CODE REVIEW

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt [hunk 1]
```diff
@@ -169,19 +171,39 @@ class OutdatedDocumentation(config: Config = Config.empty) : Rule(config) {
...
         element: KtNamedDeclaration,
-        elementDeclarationsProvider: () -> List<Declaration>
+        elementDeclarationsProvider: () -> List<Declaration>,
     ) {
         val doc = element.docComment ?: return
         val docDeclarations = getDocDeclarations(doc)
         if (docDeclarations.isNotEmpty()) {
             val elementDeclarations = elementDeclarationsProvider()
-            if (!declarationsMatch(docDeclarations, elementDeclarations)) {
+            val isInternalPrimaryConstructor = element is KtClass && isInternalOrPrivate(element.primaryConstructor)
```
Comment (schalkms) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt:180]: We call this method `reportIfDocumentationIsOutdated` with 3 different `element` types. However, in this method we check whether element has 1 specific type (e.g. `KtClass`). For me, this feels like a polymorphism code smell. See [here](https://wiki.c2.com/?SwitchStatementsSmell) for more details.

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt [hunk 2]
```diff
@@ -169,19 +171,39 @@ class OutdatedDocumentation(config: Config = Config.empty) : Rule(config) {
...
-    private fun declarationsMatch(doc: List<Declaration>, element: List<Declaration>): Boolean {
+    private fun isInternalOrPrivate(primaryConstructor: KtPrimaryConstructor?): Boolean {
+        primaryConstructor ?: return false
+        return primaryConstructor.isInternal() || primaryConstructor.isPrivate()
+    }
+
+    private fun declarationsMatch(
+        doc: List<Declaration>,
+        element: List<Declaration>,
+        shouldAllowOnlyPropertyDoc: Boolean,
```
Comment (schalkms) [inline detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt:195]: Using boolean parameters is at least controversial. One obvious reason is that it makes code harder to read and maintain.

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt [hunk 3]
```diff
@@ -151,6 +151,105 @@ class OutdatedDocumentationSpec {
...
+        fun `should not report when only public property is documented in internal constructor`() {
+            val incorrectDeclarationsOrder = """
+                /**
+                 * Exception thrown from MyLibrary that can not be created outside of the library.
+                 * @property errorCode Description of param 
+                 */
+                class MyLibraryException internal constructor(
+                    val errorCode: String,
+                    cause: Throwable?,
+                ) : Exception(cause)
```
Comment (schalkms) [inline detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt:165]: I think smaller test code snippets are sufficient here to understand what's being tested here.

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt (hunk 1 above)
Comment (atulgpt) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt:180]: Do we want me to move this check to `visitClass` method?

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt (hunk 3 above)
Comment (atulgpt) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt:165]: Is this somehow hampering the readability of the test code? I saw the other TCs which use `MyClass` but current TC is inspired by the issue https://github.com/detekt/detekt/issues/5217 also conveys why someone wants to document property in this manner

Comment (schalkms) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt:165]: Isn't shorter, more concise and less _noisy_ to add code snippets like this?

```kotlin
/**
 * Doc
 * @property b desc
*/
class A internal constructor(val b: String)
```

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt (hunk 1 above)
Comment (schalkms) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt:180]: Yes, just type `KtClass` of the 3 cases needs all those added checks.

Comment (atulgpt) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt]: I have updated the code

File: detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-rules-documentation/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentation.kt]: Removed boolean

File: detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt (hunk 3 above)
Comment (atulgpt) [reply detekt-rules-documentation/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/documentation/OutdatedDocumentationSpec.kt]: TBH even the current TC was readable to me. But I have updated the cases. Thanks for reviewing it

**Review by schalkms — APPROVED** [review]
Thanks! Looks good now.

**Pull request merged**